src/PIL/PpmImagePlugin.py

```diff
@@ -123,7 +128,176 @@ def _open(self):
+
+class PpmPlainDecoder(ImageFile.PyDecoder):
+    _pulls_fd = True
+
+    def _read_block(self, block_size=10**6):
+        return bytearray(self.fd.read(block_size))
+
+    def _find_comment_end(self, block, start=0):
+        a = block.find(b"\n", start)
+        b = block.find(b"\r", start)
+        return min(a, b) if a * b > 0 else max(a, b)  # lowest nonnegative index (or -1)
+
+    def _ignore_comments(self, block):
+        """
+        Deletes comments from block. If comment does not end in this
+        block, raises a flag.
+        """
+
+        comment_spans = False
+        while True:
+            comment_start = block.find(b"#")  # look for next comment
+            if comment_start == -1:  # no comment found
+                break
+            comment_end = self._find_comment_end(block, comment_start)
+            if comment_end != -1:  # comment ends in this block
+                block = (
+                    block[:comment_start] + block[comment_end + 1 :]
+                )  # delete comment
+            else:  # last comment continues to next block(s)
+                block = block[:comment_start]
+                comment_spans = True
+                break
+        return block, comment_spans
+
+    def _decode_bitonal(self):
+        """
+        The reason this is a separate method is that in the plain PBM
+        format all data tokens are exactly one byte, and so the
+        inter-token whitespace is optional.
+        """
+        decoded_data = bytearray()
+        total_tokens = self.size
+
+        comment_spans = False
+        tokens_read = 0
+        while True:
+            block = self._read_block()  # read next block
+            if not block:
+                raise ValueError("Reached EOF while reading data")
+
+            while block and comment_spans:
+                comment_end = self._find_comment_end(block)
+                if comment_end != -1:  # comment ends in this block
+                    comment_spans = False
+                    block = block[comment_end + 1 :]  # delete tail of previous comment
+                else:  # comment spans whole block
+                    block = self._read_block()
+
+            block, comment_spans = self._ignore_comments(block)
+
+            tokens = b"".join(block.split())
+
+            for token in tokens:
+                if token in (48, 49):
+                    tokens_read += 1
+                else:
+                    raise ValueError(f"Invalid token for this mode: {bytes([token])}")
+
+                decoded_data.append(token)
+                if tokens_read == total_tokens:  # finished!
+                    invert = bytes.maketrans(b"01", b"\xFF\x00")
+                    decoded_data = decoded_data.translate(invert)
+                    return decoded_data
+
+    def _decode_blocks(self, channels=1, depth=8):
+        decoded_data = bytearray()
+        if depth == 32:
+            maxval = 2**31 - 1  # HACK: 32-bit grayscale uses signed int
```

Comment:
I can see that your test image agrees with you. I'd like to understand why this is though. Where did you test image come from?

Reply:
If I use ImageMagick, `convert Tests/images/hopper.png -depth 32 out.pgm`,`maxval` is `2 ** 32 - 1`.

I'm also able to get the same result without explicitly setting the depth -
```
python3 -c "from PIL import Image;im = Image.open('Tests/images/hopper.png');im.convert('F').save('out.tiff')"
convert out.tiff out.pgm
cat out.pgm | head -n 3
```

Reply:
I've realised that this workaround can be avoided by using the original `maxval`. I've pushed a commit for this.

Reply:
Sorry for the late reply. It's been a while since I worked on this.
Just for the record: I don't remember how I got to the conclusion that that "hack" was needed. I think the test image I manually generated myself—so of course my hack would work ;P. I'm glad that using `maxval` makes all this avoidable.